File: webflux-dialogue/src/main/java/com/study/webflux/rag/infrastructure/monitoring/micrometer/MicrometerPipelineMetricsReporter.java
package com.study.webflux.rag.infrastructure.monitoring.micrometer;

import java.util.concurrent.TimeUnit;

import org.springframework.stereotype.Component;

import com.study.webflux.rag.application.monitoring.monitor.DialoguePipelineTracker.PipelineSummary;
import com.study.webflux.rag.application.monitoring.monitor.DialoguePipelineTracker.StageSnapshot;
import com.study.webflux.rag.application.monitoring.monitor.PipelineMetricsReporter;
import com.study.webflux.rag.domain.monitoring.model.DialoguePipelineStage;
import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Timer;

/**
 * Micrometer를 사용하여 파이프라인 메트릭을 Prometheus로 내보내는 Reporter입니다.
 */
@Component
public class MicrometerPipelineMetricsReporter implements PipelineMetricsReporter {

	private static final String METRIC_PREFIX = "dialogue.pipeline";

	private final MeterRegistry meterRegistry;

	public MicrometerPipelineMetricsReporter(MeterRegistry meterRegistry) {
		this.meterRegistry = meterRegistry;
	}

	@Override
	public void report(PipelineSummary summary) {
		recordPipelineMetrics(summary);
		recordStageMetrics(summary);
		recordStageGapMetrics(summary);
		recordLlmMetrics(summary);
		recordResponseLatencyMetrics(summary);
	}

	private void recordPipelineMetrics(PipelineSummary summary) {
		String status = summary.status().name().toLowerCase();

		// 파이프라인 실행 시간
		Timer.builder(METRIC_PREFIX + ".duration")
			.tag("status", status)
			.description("Total pipeline execution time")
			.register(meterRegistry)
			.record(summary.durationMillis(), TimeUnit.MILLISECONDS);

		// 파이프라인 실행 횟수
		Counter.builder(METRIC_PREFIX + ".executions")
			.tag("status", status)
			.description("Number of pipeline executions")
			.register(meterRegistry)
			.increment();

		// 입력 길이
		Object inputLength = summary.attributes().get("input.length");
		if (inputLength instanceof Number) {
			meterRegistry.summary(METRIC_PREFIX + ".input.length")
				.record(((Number) inputLength).doubleValue());
		}
	}

	private void recordStageMetrics(PipelineSummary summary) {
		for (StageSnapshot stage : summary.stages()) {
			String stageName = stage.stage().name().toLowerCase();
			String stageStatus = stage.status().name().toLowerCase();

			// 스테이지 실행 시간
			if (stage.durationMillis() >= 0) {
				Timer.builder(METRIC_PREFIX + ".stage.duration")
					.tag("stage", stageName)
					.tag("status", stageStatus)
					.description("Stage execution time")
					.register(meterRegistry)
					.record(stage.durationMillis(), TimeUnit.MILLISECONDS);
			}

			// 스테이지별 속성 메트릭
			recordStageAttributes(stage);
		}
	}

	private void recordStageAttributes(StageSnapshot stage) {
		String stageName = stage.stage().name().toLowerCase();

		// MEMORY_RETRIEVAL: 검색된 메모리 수
		if (stage.stage() == DialoguePipelineStage.MEMORY_RETRIEVAL) {
			recordCounterFromAttribute(stage,
				"memory.count",
				METRIC_PREFIX + ".memory.retrieved",
				stageName);

			// RAG 품질 메트릭 (Phase 1B)
			recordRagQualityMetrics(stage);
		}

		// RETRIEVAL: 검색된 문서 수
		if (stage.stage() == DialoguePipelineStage.RETRIEVAL) {
			recordCounterFromAttribute(stage,
				"document.count",
				METRIC_PREFIX + ".documents.retrieved",
				stageName);
		}

		// SENTENCE_ASSEMBLY: 생성된 문장 수
		if (stage.stage() == DialoguePipelineStage.SENTENCE_ASSEMBLY) {
			recordCounterFromAttribute(stage,
				"sentence.count",
				METRIC_PREFIX + ".sentences.generated",
				stageName);
		}

		// TTS_SYNTHESIS: 오디오 청크 수
		if (stage.stage() == DialoguePipelineStage.TTS_SYNTHESIS) {
			recordCounterFromAttribute(stage,
				"audio.chunks",
				METRIC_PREFIX + ".audio.chunks",
				stageName);
		}
	}

	/**
	 * RAG 품질 메트릭을 기록합니다. Phase 1B: RAG Quality Monitoring
	 */
	private void recordRagQualityMetrics(StageSnapshot stage) {
		// Memory count by type
		Object experientialCount = stage.attributes().get("memory.experiential.count");
		if (experientialCount instanceof Number) {
			meterRegistry.gauge("rag.memory.count",
				io.micrometer.core.instrument.Tags.of("memory_type", "experiential"),
				((Number) experientialCount).doubleValue());
		}

		Object factualCount = stage.attributes().get("memory.factual.count");
		if (factualCount instanceof Number) {
			meterRegistry.gauge("rag.memory.count",
				io.micrometer.core.instrument.Tags.of("memory_type", "factual"),
				((Number) factualCount).doubleValue());
		}
	}

	private void recordCounterFromAttribute(StageSnapshot stage,
		String attrKey,
		String metricName,
		String stageName) {
		Object value = stage.attributes().get(attrKey);
		if (value instanceof Number) {
			meterRegistry.summary(metricName)
				.record(((Number) value).doubleValue());
		}
	}

	private void recordLlmMetrics(PipelineSummary summary) {
		for (StageSnapshot stage : summary.stages()) {
			if (stage.stage() != DialoguePipelineStage.LLM_COMPLETION) {
				continue;
			}

			// 모델명
			Object model = stage.attributes().get("model");
			String modelTag = model != null ? model.toString() : "unknown";

			// 토큰 사용량
			recordTokenCounter(stage, "prompt.tokens", "prompt", modelTag);
			recordTokenCounter(stage, "completion.tokens", "completion", modelTag);
			recordTokenCounter(stage, "total.tokens", "total", modelTag);

			// 비용 (있는 경우)
			Object cost = stage.attributes().get("cost.usd");
			if (cost instanceof Number) {
				meterRegistry.gauge("llm.cost.usd", ((Number) cost).doubleValue());
			}
		}
	}

	private void recordTokenCounter(StageSnapshot stage,
		String attrKey,
		String tokenType,
		String model) {
		Object value = stage.attributes().get(attrKey);
		if (value instanceof Number) {
			Counter.builder("llm.tokens")
				.tag("type", tokenType)
				.tag("model", model)
				.description("LLM token usage")
				.register(meterRegistry)
				.increment(((Number) value).doubleValue());
		}
	}

	private void recordResponseLatencyMetrics(PipelineSummary summary) {
		// 첫 응답 지연 시간 (TTFB)
		if (summary.firstResponseLatencyMillis() != null) {
			Timer.builder(METRIC_PREFIX + ".response.first")
				.description("Time to first response")
				.register(meterRegistry)
				.record(summary.firstResponseLatencyMillis(), TimeUnit.MILLISECONDS);
		}

		// 마지막 응답 지연 시간 (TTLB)
		if (summary.lastResponseLatencyMillis() != null) {
			Timer.builder(METRIC_PREFIX + ".response.last")
				.description("Time to last response")
				.register(meterRegistry)
				.record(summary.lastResponseLatencyMillis(), TimeUnit.MILLISECONDS);
		}
	}

	/**
	 * Stage 간 Gap (전환 시간)을 기록합니다. Phase 1A: Pipeline Bottleneck Analysis
	 */
	private void recordStageGapMetrics(PipelineSummary summary) {
		var stages = summary.stages().stream()
			.sorted((a, b) -> {
				if (a.startedAt() == null)
					return 1;
				if (b.startedAt() == null)
					return -1;
				return a.startedAt().compareTo(b.startedAt());
			})
			.toList();

		for (int i = 0; i < stages.size() - 1; i++) {
			StageSnapshot current = stages.get(i);
			StageSnapshot next = stages.get(i + 1);

			// Gap 계산: 현재 Stage 종료 → 다음 Stage 시작
			if (current.finishedAt() != null && next.startedAt() != null) {
				long gapMillis = java.time.Duration.between(
					current.finishedAt(),
					next.startedAt()).toMillis();

				// Gap이 음수면 병렬 실행된 것이므로 0으로 처리
				if (gapMillis >= 0) {
					Timer.builder(METRIC_PREFIX + ".stage.gap")
						.tag("from_stage", current.stage().name().toLowerCase())
						.tag("to_stage", next.stage().name().toLowerCase())
						.description("Time gap between stage transitions")
						.register(meterRegistry)
						.record(gapMillis, TimeUnit.MILLISECONDS);
				}
			}
		}
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/infrastructure/monitoring/micrometer/MicrometerPipelineMetricsReporter.kt
package com.miyou.app.infrastructure.monitoring.micrometer

import com.miyou.app.domain.monitoring.model.DialoguePipelineStage
import com.miyou.app.domain.monitoring.model.PipelineStatus
import com.miyou.app.domain.monitoring.model.PipelineSummary
import com.miyou.app.domain.monitoring.model.StageSnapshot
import com.miyou.app.domain.monitoring.model.StageStatus
import com.miyou.app.domain.monitoring.port.PipelineMetricsReporter
import com.miyou.app.infrastructure.monitoring.config.CostTrackingMetricsConfiguration
import com.miyou.app.infrastructure.monitoring.config.LlmMetricsConfiguration
import com.miyou.app.infrastructure.monitoring.config.UxMetricsConfiguration
import io.micrometer.core.instrument.Counter
import io.micrometer.core.instrument.MeterRegistry
import io.micrometer.core.instrument.Tags
import io.micrometer.core.instrument.Timer
import org.springframework.stereotype.Component
import java.time.Duration
import java.util.concurrent.TimeUnit

/**
 * Micrometer 기반 파이프라인 메트릭 리포터.
 */
@Component
class MicrometerPipelineMetricsReporter(
    private val meterRegistry: MeterRegistry,
    private val llmMetrics: LlmMetricsConfiguration,
    private val costMetrics: CostTrackingMetricsConfiguration,
    private val uxMetrics: UxMetricsConfiguration,
) : PipelineMetricsReporter {
    override fun report(summary: PipelineSummary) {
        recordPipelineMetrics(summary)
        recordStageMetrics(summary)
        recordStageGapMetrics(summary)
        recordLlmMetrics(summary)
        recordResponseLatencyMetrics(summary)
        recordCostMetrics(summary)
        recordUxMetrics(summary)
    }

    private fun recordPipelineMetrics(summary: PipelineSummary) {
        val status = summary.status.name.lowercase()

        Timer
            .builder(METRIC_PREFIX + ".duration")
            .tag("status", status)
            .description("Total pipeline execution time")
            .register(meterRegistry)
            .record(summary.durationMillis(), TimeUnit.MILLISECONDS)

        Counter
            .builder(METRIC_PREFIX + ".executions")
            .tag("status", status)
            .description("Number of pipeline executions")
            .register(meterRegistry)
            .increment()

        val inputLength = summary.attributes["input.length"]
        if (inputLength is Number) {
            meterRegistry
                .summary(METRIC_PREFIX + ".input.length")
                .record(inputLength.toDouble())
        }
    }

    private fun recordStageMetrics(summary: PipelineSummary) {
        summary.stages.forEach { stage ->
            val stageName = stage.stage.name.lowercase()
            val stageStatus = stage.status.name.lowercase()

            if (stage.durationMillis >= 0) {
                Timer
                    .builder(METRIC_PREFIX + ".stage.duration")
                    .tag("stage", stageName)
                    .tag("status", stageStatus)
                    .description("Stage execution time")
                    .register(meterRegistry)
                    .record(stage.durationMillis, TimeUnit.MILLISECONDS)
            }

            recordStageAttributes(stage)
        }
    }

    private fun recordStageAttributes(stage: StageSnapshot) {
        val stageName = stage.stage.name.lowercase()
        when (stage.stage) {
            DialoguePipelineStage.MEMORY_RETRIEVAL -> {
                recordCounterFromAttribute(
                    stage,
                    "memory.count",
                    METRIC_PREFIX + ".memory.retrieved",
                    stageName,
                )
                recordRagQualityMetrics(stage)
            }

            DialoguePipelineStage.RETRIEVAL -> {
                recordCounterFromAttribute(
                    stage,
                    "document.count",
                    METRIC_PREFIX + ".documents.retrieved",
                    stageName,
                )
            }

            DialoguePipelineStage.SENTENCE_ASSEMBLY -> {
                recordCounterFromAttribute(
                    stage,
                    "sentence.count",
                    METRIC_PREFIX + ".sentences.generated",
                    stageName,
                )
            }

            DialoguePipelineStage.TTS_SYNTHESIS -> {
                recordCounterFromAttribute(
                    stage,
                    "audio.chunks",
                    METRIC_PREFIX + ".audio.chunks",
                    stageName,
                )
            }

            else -> {}
        }
    }

    private fun recordRagQualityMetrics(stage: StageSnapshot) {
        val experientialCount = stage.attributes["memory.experiential.count"]
        if (experientialCount is Number) {
            meterRegistry.gauge(
                "rag.memory.count",
                Tags.of("memory_type", "experiential"),
                experientialCount.toDouble(),
            )
        }

        val factualCount = stage.attributes["memory.factual.count"]
        if (factualCount is Number) {
            meterRegistry.gauge(
                "rag.memory.count",
                Tags.of("memory_type", "factual"),
                factualCount.toDouble(),
            )
        }
    }

    private fun recordCounterFromAttribute(
        stage: StageSnapshot,
        attrKey: String,
        metricName: String,
        @Suppress("UNUSED_PARAMETER") stageName: String,
    ) {
        val value = stage.attributes[attrKey]
        if (value is Number) {
            meterRegistry
                .summary(metricName)
                .record(value.toDouble())
        }
    }

    private fun recordLlmMetrics(summary: PipelineSummary) {
        summary.stages
            .filter { it.stage == DialoguePipelineStage.LLM_COMPLETION }
            .forEach { stage ->
                val model = stage.attributes["model"]
                val modelTag = model?.toString() ?: "unknown"

                llmMetrics.recordLlmRequest()

                when (stage.status) {
                    StageStatus.COMPLETED -> {
                        llmMetrics.recordLlmSuccess(modelTag)
                    }

                    StageStatus.FAILED -> {
                        val error = stage.attributes["error.type"]
                        val errorType = error?.toString() ?: "unknown"
                        llmMetrics.recordLlmFailure(modelTag, errorType)
                    }

                    else -> {}
                }

                recordTokenCounter(stage, "prompt.tokens", "prompt", modelTag)
                recordTokenCounter(stage, "completion.tokens", "completion", modelTag)
                recordTokenCounter(stage, "total.tokens", "total", modelTag)

                val promptTokens = stage.attributes["prompt.tokens"]
                if (promptTokens is Number) {
                    llmMetrics.recordPromptLength(promptTokens.toInt())
                }

                val completionTokens = stage.attributes["completion.tokens"]
                if (completionTokens is Number) {
                    llmMetrics.recordCompletionLength(completionTokens.toInt())
                }

                if (stage.durationMillis >= 0) {
                    llmMetrics.recordResponseTime(stage.durationMillis)
                    llmMetrics.recordResponseTimeByModel(modelTag, stage.durationMillis)
                }

                val cost = stage.attributes["cost.usd"]
                if (cost is Number) {
                    meterRegistry.gauge("llm.cost.usd", cost.toDouble())
                }
            }
    }

    private fun recordTokenCounter(
        stage: StageSnapshot,
        attrKey: String,
        tokenType: String,
        model: String,
    ) {
        val value = stage.attributes[attrKey]
        if (value is Number) {
            Counter
                .builder("llm.tokens")
                .tag("type", tokenType)
                .tag("model", model)
                .description("LLM token usage")
                .register(meterRegistry)
                .increment(value.toDouble())
        }
    }

    private fun recordResponseLatencyMetrics(summary: PipelineSummary) {
        summary.firstResponseLatencyMillis?.let { firstLatency ->
            Timer
                .builder(METRIC_PREFIX + ".response.first")
                .description("Time to first response")
                .register(meterRegistry)
                .record(firstLatency, TimeUnit.MILLISECONDS)
        }

        summary.lastResponseLatencyMillis?.let { lastLatency ->
            Timer
                .builder(METRIC_PREFIX + ".response.last")
                .description("Time to last response")
                .register(meterRegistry)
                .record(lastLatency, TimeUnit.MILLISECONDS)
        }
    }

    private fun recordStageGapMetrics(summary: PipelineSummary) {
        val stages =
            summary.stages.sortedWith { a, b ->
                when {
                    a.startedAt == null -> 1
                    b.startedAt == null -> -1
                    else -> a.startedAt.compareTo(b.startedAt)
                }
            }

        for (i in 0 until stages.size - 1) {
            val current = stages[i]
            val next = stages[i + 1]
            if (current.finishedAt != null && next.startedAt != null) {
                val gapMillis = Duration.between(current.finishedAt, next.startedAt).toMillis()
                if (gapMillis >= 0) {
                    Timer
                        .builder(METRIC_PREFIX + ".stage.gap")
                        .tag("from_stage", current.stage.name.lowercase())
                        .tag("to_stage", next.stage.name.lowercase())
                        .description("Time gap between stage transitions")
                        .register(meterRegistry)
                        .record(gapMillis, TimeUnit.MILLISECONDS)
                }
            }
        }
    }

    private fun recordCostMetrics(summary: PipelineSummary) {
        summary.stages.forEach { stage ->
            when (stage.stage) {
                DialoguePipelineStage.LLM_COMPLETION -> {
                    val model = stage.attributes["model"]?.toString() ?: "unknown"
                    val promptTokens = stage.attributes["prompt.tokens"]
                    val completionTokens = stage.attributes["completion.tokens"]

                    if (promptTokens is Number && completionTokens is Number) {
                        val promptTokenValue = promptTokens.toInt()
                        val completionTokenValue = completionTokens.toInt()
                        costMetrics.recordLlmCost(model, promptTokenValue, completionTokenValue)

                        val userIdObj = summary.attributes["user.id"]
                        if (userIdObj != null) {
                            val userId = userIdObj.toString()
                            val cost =
                                calculateLlmCost(
                                    model,
                                    promptTokenValue,
                                    completionTokenValue,
                                )
                            costMetrics.recordUserLlmCost(userId, model, cost)
                        }
                    }
                }

                DialoguePipelineStage.TTS_SYNTHESIS -> {
                    val provider = stage.attributes["provider"]?.toString() ?: "unknown"
                    val characters = stage.attributes["characters"]
                    if (characters is Number) {
                        costMetrics.recordTtsCost(provider, characters.toInt())
                    }
                }

                else -> {}
            }
        }
    }

    private fun recordUxMetrics(summary: PipelineSummary) {
        summary.firstResponseLatencyMillis
            ?.takeIf { it >= 0 }
            ?.let { uxMetrics.recordFirstResponseLatency(it) }

        val totalDurationMs = summary.durationMillis()
        if (totalDurationMs >= 0) {
            uxMetrics.recordCompleteResponseLatency(totalDurationMs)
        }

        if (summary.status == PipelineStatus.FAILED) {
            val errorTypeObj = summary.attributes["error.type"]
            val errorType = errorTypeObj?.toString() ?: "unknown"
            uxMetrics.recordError(errorType)
        }
    }

    private fun calculateLlmCost(
        model: String,
        promptTokens: Int,
        completionTokens: Int,
    ): Double {
        val modelLower = model.lowercase()
        return when {
            modelLower.contains("gpt-4o") && !modelLower.contains("mini") -> {
                (promptTokens / 1_000_000.0) * 2.50 + (completionTokens / 1_000_000.0) * 10.00
            }

            modelLower.contains("gpt-4o-mini") -> {
                (promptTokens / 1_000_000.0) * 0.150 + (completionTokens / 1_000_000.0) * 0.600
            }

            modelLower.contains("gpt-4-turbo") -> {
                (promptTokens / 1_000_000.0) * 10.00 + (completionTokens / 1_000_000.0) * 30.00
            }

            modelLower.contains("gpt-3.5-turbo") -> {
                (promptTokens / 1_000_000.0) * 0.50 + (completionTokens / 1_000_000.0) * 1.50
            }

            else -> {
                ((promptTokens + completionTokens) / 1_000_000.0) * 5.0
            }
        }
    }

    companion object {
        private const val METRIC_PREFIX = "dialogue.pipeline"
    }
}
